pkg/etcd/init/init.go

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package init
+
+import (
+	"context"
+	"fmt"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"github.com/cilium/cilium/pkg/kvstore"
+)
+
+const (
+	rootRoleName = "root"
+)
+
+func InitEtcd(ctx context.Context, client *clientv3.Client, clusterName string) error {
+	_, err := client.Put(ctx, kvstore.HasClusterConfigPath, "true")
+	if err != nil {
+		return err
+	}
+	err = setupRootUser(ctx, client)
+	if err != nil {
+		return err
+	}
+	err = setupAdminUser(ctx, client, clusterName)
+	if err != nil {
+		return err
+	}
+	err = setupExternalWorkloadUser(ctx, client)
+	if err != nil {
+		return err
+	}
+	err = setupRemoteUser(ctx, client)
+	if err != nil {
+		return err
+	}
+	_, err = client.AuthEnable(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setupRootUser(ctx context.Context, client *clientv3.Client) error {
+	rootUsername := "root"
+	_, err := client.UserAddWithOptions(ctx, rootUsername, "", &clientv3.UserAddOptions{NoPassword: true})
+	if err != nil {
+		return err
+	}
+	_, err = client.UserGrantRole(ctx, rootUsername, rootRoleName)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func setupAdminUser(ctx context.Context, client *clientv3.Client, clusterName string) error {
+	adminUsername := fmt.Sprintf("admin-%s", clusterName)
+	_, err := client.UserAddWithOptions(ctx, adminUsername, "", &clientv3.UserAddOptions{NoPassword: true})
+	if err != nil {
+		return err
+	}
+	_, err = client.UserGrantRole(ctx, adminUsername, rootRoleName)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func setupExternalWorkloadUser(ctx context.Context, client *clientv3.Client) error {
+	externalWorkloadUsername := "externalworkload"
+	externalWorkloadRoleName := "externalworkload"
+	_, err := client.UserAddWithOptions(ctx, externalWorkloadUsername, "", &clientv3.UserAddOptions{NoPassword: true})
+	if err != nil {
+		return err
+	}
+	_, err = client.RoleAdd(ctx, externalWorkloadRoleName)
+	if err != nil {
+		return err
+	}
+	_, err = client.UserGrantRole(ctx, externalWorkloadUsername, externalWorkloadRoleName)
+	if err != nil {
+		return err
+	}
+	_, err = client.RoleGrantPermission(ctx, externalWorkloadRoleName, "\x00", "\x00", clientv3.PermissionType(clientv3.PermRead))
+	if err != nil {
+		return err
+	}
+	_, err = client.RoleGrantPermission(ctx, externalWorkloadRoleName, "cilium/state/noderegister/v1/", "cilium/state/noderegister/v10", clientv3.PermissionType(clientv3.PermReadWrite))
```

Comment:
```suggestion
	_, err = client.RoleGrantPermission(ctx, externalWorkloadRoleName, "cilium/state/noderegister/v1/", clientv3.GetPrefixRangeEnd("cilium/state/noderegister/v1/"), clientv3.PermissionType(clientv3.PermReadWrite))
```

Same for the one below.
See https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/role_command.go#L231